### PR TITLE
Feature/task1 automation users api

### PR DESCRIPTION
**Pull request by yergu**
** To Reproduce **
1.Pull the code on your local machine.
2.When inside the repo, navigate to Automation\usersAPI
3.You will need the Service Plan Id, Service Principal Password, Tenant Id and Subscription Id to run this script. The required information can be received from the key vault found in Azure.
4.Switch your default shell to power shell (if you are running this from within VS code).
5.The script can be run using the following command (while inside the usersAPI folder) change the {} with the the parameter value
6 ./deploy-users-templet.ps1 -subscriptionId {} -tenantId {} -servicePrincipalId {} -servicePrincipalPassword {} -resourceGroupName {} -functionName {} -storageAccountName {}-appServicePlanName {} -location {} -templateFilePath  {}

**Expected behavior**
if it deployed successfully , when you go to azure Portal you will see new resource group created

### CODE REVIEW

File: automation/usersAPI/template.json [hunk 1]
```diff
@@ -0,0 +1,363 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "functionName": {
+            "type": "tring"
```
Comment (jinudean) [inline automation/usersAPI/template.json:6]: syntax error "type":"string"

Comment (yergu) [reply automation/usersAPI/template.json:6]: I will fix it , thanks

Comment (yergu) [reply automation/usersAPI/template.json:6]: I fixed it

**Review by ahmadafrasiyab — APPROVED** [review]
i was able to run and get the resources deployed. The errors I got are the same as my own script. Other than that it runs ok and deploys three resources. The parameters file needs to be deleted

File: automation/usersAPI/parameters.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        
+        "sites_nsc_fun_dev_usw2_tsmtestfunction_name": {
```
[the remark this answers is not in the document]
Comment (yergu) [reply automation/usersAPI/parameters.json]: I removed the paramater.json

File: automation/usersAPI/deploy-users-templet.ps1 [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+$credentials = New-Object -TypeName System.Management.Automation.PSCredential($servicePrincipalId, $securePassword);
+
+
+Connect-AzAccount -Credential $credentials -ServicePrincipal -Tenant $TenantId -Subscription $SubscriptionId;
+
+Set-AzContext $SubscriptionId
+New-AzResourceGroup -Name $resourceGroupName -Location $location
+
+New-AzResourceGroupDeployment -resourceGroupName $resourceGroupName -functionName $functionName -storageAccountName $storageAccountName -AppServicePlanName  $AppServicePlanName -location $location -TemplatePath $TemplatePath
```
[the remark this answers is not in the document]
Comment (yergu) [reply automation/usersAPI/deploy-users-templet.ps1]: I run the code and the resources deployed successfully but I got the Xml error like you

**Comment by yergu** [comment]
I removed the parameter json file

**Comment by ahmadafrasiyab** [comment]
I have run it successfully this time. At this point, yours and my script results are identical. The errors we get are identical and I believe it is to do with the template format that has some issue that both of us have been unable to identify. @jinudean template file was much shorter than ours and doesn't return those errors so I am guessing that the initial template we took from Azure has some problems. In any case, I see that the resources are deployed as follows:
image

Also, I see that you have gotten rid of the parameters.json file. I approve this PR.

**Review by ahmadafrasiyab — APPROVED** [review]
Approved with feedback posted previously.

File: automation/usersAPI/deploy-users-templet.ps1 [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+Clear-AzContext -Force;
+$securePassword = $servicePrincipalPassword | ConvertTo-SecureString -AsPlainText -Force
+
+$credentials = New-Object -TypeName System.Management.Automation.PSCredential($servicePrincipalId, $securePassword);
+
+
+Connect-AzAccount -Credential $credentials -ServicePrincipal -Tenant $tenantId -Subscription $subscriptionId;
+
+Set-AzContext $SubscriptionId
+New-AzResourceGroup -Name $resourceGroupName -Location $location
```
Comment (theanticrumpet) [inline automation/usersAPI/deploy-users-templet.ps1:61]: What happens if the Resource Group already exists?

File: automation/usersAPI/deploy-users-templet.ps1 [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    $location,
+
+    [Parameter(Mandatory=$True)]
+    $templateFilePath
+)
+
+#open log
+$prefix = "yergu-Arm-deplo"
+$stamp = (Get-Date).toString().Replace("/","-").Replace(":","-")
+Start-Transcript "./$prefix-$stamp.log"
```
Comment (theanticrumpet) [inline automation/usersAPI/deploy-users-templet.ps1:49]: I like what you did here to get a timestamp for the log file name but is there a better way to format this than doing a match and replace?

**Review by theanticrumpet — APPROVED** [review]
I tested out your code here. Some more details about what is expected for the params (or validation) would be appreciated.

So the code did successfully create the resource group and populate it with a function, service plan, and storage account:
image

BUT it ran with a couple of errors:
image
(I'll put the full log file in the private slack channel, not sure what all in it shouldn't be posted to a public repo)

I'll go ahead and give an approve for the sprint, especially since this work won't directly interfere with anyone else's sprint 2 work. But while I don't know exactly what those errors mean off the top of my head they are concerning.

File: automation/usersAPI/deploy-users-templet.ps1 (hunk 4 above)
Comment (yergu) [reply automation/usersAPI/deploy-users-templet.ps1:61]: It says resource group already exist and prompts me if I want to update it or not

File: automation/usersAPI/deploy-users-templet.ps1 (hunk 5 above)
Comment (yergu) [reply automation/usersAPI/deploy-users-templet.ps1:49]: I am not sure I will figure out

File: automation/usersAPI/deploy-users-templet.ps1 [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+Clear-AzContext -Force;
+$securePassword = $servicePrincipalPassword | ConvertTo-SecureString -AsPlainText -Force
+
+$credentials = New-Object -TypeName System.Management.Automation.PSCredential($servicePrincipalId, $securePassword);
+
+
+Connect-AzAccount -Credential $credentials -ServicePrincipal -Tenant $tenantId -Subscription $subscriptionId;
+
+Set-AzContext $SubscriptionId
```
Comment (toddysm) [inline automation/usersAPI/deploy-users-templet.ps1:60]: This is not needed. It is already in the `Connect-AzAccount`

**Pull request merged**